### PR TITLE
Exclude marketing channels in "Installed extensions" and "Discover more marketing tools" cards

### LINKED ISSUE #34907: Modify "Installed extensions" and "Discover more marketing tools" to exclude channels
This is a task issue to modify the "Installed extensions" card and "Discover more marketing tools" card to exclude marketing channels in the multichannel marketing page. Marketing channels would show up in the "Channels" card as per task issue https://github.com/woocommerce/woocommerce/issues/34906.

Link to Figma design: 7cNed5RvTasoLgn38Hea0R-fi-2145%3A84807

image

The design is the same as what we already have now since previous task issue https://github.com/woocommerce/woocommerce/issues/34025. What we need to do in this issue is to change the data fetching and handling logic.

### PR DESCRIPTION
**Pull request by ecgan**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/34907.

In this PR, we filter out marketing channels in the "Installed extensions" and "Discover more marketing tools" cards. This is because marketing channels already show up in its own "Channels" card.

As a result, there should be no more "Sales channels" tab in the "Discover more marketing tools" card. Each extension should only appear in one card.

Screenshot:

image

### How to test the changes in this Pull Request:

- [x] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

Pre-condition: 

- Enable WooCommerce Multichannel Marketing experience in WooCommerce Settings > Advanced > Features (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`).

Test instructions:

1. Go to Marketing page at `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`.
2. Each extension should only appear in one card.
3. There should be no "Sales channel" tab in the "Discover more marketing tools" card.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [x] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by ecgan** [comment]
I have rebased the commits onto trunk branch, and changed this PR's base branch from `feature/34904-marketing-introduction-banner` to `trunk`, so that this PR does not depend on other PR and is ready for review.

File: plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts]: From the perspective of this hook, it's named `useInstalledPlugins` but filters out some plugins internally, which might surprise the hook users.

While this can be clarified by this code comment, it might be possible to rename this hook, or to add hook parameters to indicate whether filtering out the registered and/or recommended ones.

Comment (eason9487) [reply plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts]: Another idea is to keep exposing the original `installedPlugins` and also exposing `installedPluginsWithoutChannels` separately.

File: plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts]: Since `installedPlugins` are hydrated directly into the wp-data store via PHP preload data, the results of `useRegisteredChannels` and `useRecommendedChannels` will definitely be later than it. This causes the `installedPluginsWithoutChannels` to have incorrect data for a moment to a few seconds.

Although there is currently an additional loading state check in the location where this hook is used, which won't cause other problems in the runtime results, it's still recommended to fix it.

Demo a potential problem of flickering UI when data is not yet prepared:

https://user-images.githubusercontent.com/17420811/226311807-746c8159-4f39-4f85-a7c9-4f71bc5b4418.mp4

<br />

Demo code:

```ts
import '~/marketing/data';
import '~/marketing/data-multichannel';
import { InstalledExtensions } from './InstalledExtensions';
import './MarketingOverviewMultichannel.scss';

export const MarketingOverviewMultichannel: React.FC = () => {
	return (
		<div className="woocommerce-marketing-overview-multichannel">
			<InstalledExtensions />
		</div>
	);
};
```

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPlugins.ts
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPlugins.ts:1]: This hook seems to have the same two concerns/issues as the `useInstalledPlugins`.

To make it easier to simulate/reproduce the problem of incorrect data, steps I tried:

```js
import { __unstableAwaitPromise } from '@wordpress/data-controls';

export function* getRecommendedChannels() {
	// Delay the resolver.
	yield __unstableAwaitPromise( new Promise( ( resolve ) => setTimeout( resolve, 3000 ) ) );
```

```ts
import { useRecommendedPlugins } from './DiscoverTools/useRecommendedPlugins';

export const MarketingOverviewMultichannel: React.FC = () => {
	const { plugins } = useRecommendedPlugins();
	console.log( 'plugins', plugins );
```


image

**Review by eason9487 — COMMENTED** [review]
Thanks for the adjustment.

My main concerns are about two hooks may return incorrect data at certain times. Although the current runtime results on UI are shown correctly, it is still recommended to fix it first.

The two remaining ones about naming or usability are not blockers but are also recommended to adjust.

File: plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts
Comment (ecgan) [reply plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts]: > From the perspective of this hook, it's named `useInstalledPlugins` but filters out some plugins internally, which might surprise the hook users.
> 
> While this can be clarified by this code comment, it might be possible to rename this hook, or to add hook parameters to indicate whether filtering out the registered and/or recommended ones.

You got a point. I called it `useInstalledPlugins` because we uses the classic marketing wp.data store "wc/marketing" and it uses the term "plugin". I'm thinking I can rename them to `useInstalledExtensions` instead, because it is meant to be used with the "Installed extensions" card. Similarly, I'm thinking to rename `useRecommendedPlugins` to `useDiscoverTools`.

> Another idea is to keep exposing the original **`installedPlugins`** and also exposing `installedPluginsWithoutChannels` separately.

I less prefer this idea because we won't be using the `installedPlugins`.

Comment (eason9487) [reply plugins/woocommerce-admin/client/marketing/hooks/useInstalledPlugins.ts]: > I'm thinking I can rename them to `useInstalledExtensions` instead, because it is meant to be used with the "Installed extensions" card.

Renaming to `useInstalledExtensions` perhaps makes it more concrete in association with the "Installed extensions" card but it seems to still have the same confusion in the `useInstalledPlugins` naming: The `const { installedPlugins } = useInstalledExtensions()` statement still doesn't look like it will filter out some plugins.

**Comment by ecgan** [comment]
@eason9487 , thanks for the sharp eyes! And the demo code, video and screenshot, really appreciate those! 🙂 🙏 

I made quite some commits (some back and forth; naming is hard 😂 ) to address the problems. Here's a summary of the changes:

- `useInstalledPlugins` hook:
    - I renamed `useInstalledPlugins` hook to `useInstalledPluginsWithoutChannels`. Although the hook name is long, I think this shows the clearest meaning.
    - The `installedPlugins` property in the returned object is renamed to `data`. I change this to make it consistent with other hooks.
    - When loading is in progress, `data` will be an empty array. It returns the final correct data when loading is completed. So there should be no intermediate incorrect data.

- `useRecommendedPlugins` hook:
    - I renamed `useRecommendedPlugins` to `useRecommendedPluginsWithoutChannels`.
    - The `plugins` property in the returned object is renamed to `data`.
    - When loading is in progress, `data` will be an empty array. It returns the final correct data when loading is completed. So there should be no intermediate incorrect data.

This is now ready for another round of review. 🙏

File: plugins/woocommerce-admin/client/marketing/hooks/UseInstalledPluginsWithoutChannels.ts
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/hooks/UseInstalledPluginsWithoutChannels.ts:1]: This filename was missed to start with a lowercase.

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/DiscoverTools.test.tsx
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/DiscoverTools.test.tsx]: This test suite mocks four non-existent functions `loadInstalledExtensionsAfterActivation` and it can still pass the tests. Looks like it doesn't need to mock the return value of function in each test?

Simply returning an empty object on the top of the test should be ok:

```js
jest.mock( '~/marketing/hooks', () => ( {
	useInstalledPluginsWithoutChannels: jest.fn( () => ( {} ) ),
} ) );
```

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/PluginsTabPanel.tsx
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/PluginsTabPanel.tsx]: Is there a particular reason to rename this function or was is missed to further rename to the latest name?

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts
Comment (eason9487) [inline plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts:70]: If API will never return an empty array as `dataRecommendedPlugins`, this logic should be fine, and it would be a bit nicer to note this here.

But if any case the `dataRecommendedPlugins` was an empty array, the `isInitializing` would not fit with its description:

https://github.com/woocommerce/woocommerce/blob/e281c09bc67eb3b94fde3bf82a93f358ea2b2ae3/plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts#L15-L18

**Review by eason9487 — COMMENTED** [review]
Thanks for the update.

Most of the code looks good. The only needed tweak is the filename `UseInstalledPluginsWithoutChannels.ts` doesn't follow the convention of React hook. The rest comments are minor things or a question.

I guess this repo is ticked the "Dismiss stale pull request approvals when new commits are pushed" option for the `trunk` branch, so even though I would like to approve it in this round but probably it would still need a new approval after pushing the file renaming commit. So feel free to ping me afterward.

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/DiscoverTools.test.tsx
Comment (ecgan) [reply plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/DiscoverTools.test.tsx]: Thanks! Fixed in [`b18e170` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/b18e1705cd9c32274a1ab6ed5af70010ee04f381).

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/PluginsTabPanel.tsx
Comment (ecgan) [reply plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/PluginsTabPanel.tsx]: Thanks, missed this when I doing the renaming. Fixed in [`5e6cd11` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/5e6cd11028fe6219c264de067b1550c6f097903a).

File: plugins/woocommerce-admin/client/marketing/hooks/UseInstalledPluginsWithoutChannels.ts
Comment (ecgan) [reply plugins/woocommerce-admin/client/marketing/hooks/UseInstalledPluginsWithoutChannels.ts:1]: This is fixed, the result is in [`7637977` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/76379770a184a307bb091b9b9d9eef952a365888).

File: plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts
Comment (ecgan) [reply plugins/woocommerce-admin/client/marketing/overview-multichannel/DiscoverTools/useRecommendedPluginsWithoutChannels.ts:70]: Thanks. The API data may be an empty array. I have changed the description in [`0149252` (#37126)](https://github.com/woocommerce/woocommerce/pull/37126/commits/01492520d326232c98806ed37ae90b6b7b95c73e).

**Comment by ecgan** [comment]
@eason9487 , thanks, this is ready for another round. 🙂 🙏

**Review by eason9487 — APPROVED** [review]
LGTM.

**Pull request merged**
